euphonic/util.py

```diff
@@ -237,6 +238,28 @@ def custom_decode(dct):
             if isinstance(value, (float, complex))}
 
 
+def mode_gradients_to_widths(mode_gradients: Quantity, cell_vectors: Quantity
```

Comment:
This function logic is actually taken from the CASTEP implementation, as I wanted to get a close a comparison as possible to the CASTEP reference data. In CASTEP itself it only uses this version if the points have been supplied by `phonon_fine_kpoints_list` rather than `phonon_fine_mp_grid`, otherwise it uses a different estimation based on the grid.

Now that we know the adaptive broadening 'works', I don't mind using a different estimation. But in the end, I don't think the scaling of the mode widths matters too much, it depends really on how much broadening you want... I guess the important thing is this function gives a sensible estimate.

> I think it might be safer to determine the characteristic q-point distance before the DOS is calculated, and then pass that information along to where it is needed later

But it is? This is called in `force_constants.py` just before returning the mode widths. Unless it would be better to return the mode gradients from that calculation and let users do their own conversion (or use this function if they wish)? The main reason I moved this into its own function was because `.phonon_dos` includes the mode gradients, not mode widths, and I needed an easy way to convert for testing.

Reply:
I see. It wasn't entirely obvious where the function is used because it's in `util.py`. In that case maybe it is enough to state the assumption in the docstring: e.g. "The set of calculated mode_gradients should not be symmetry-reduced; the density of q-point sampling is estimated using the size of this array."

Reply:
Done!